engines/router/missionctl/handlers/batch_http_handler.go

```diff
@@ -0,0 +1,124 @@
+	ErrorMsg   string      `json:"error,omitempty"`
+	Data       interface{} `json:"data,omitempty"`
+}
+
+type batchResult struct {
+	BatchResult []batchResponse `json:"batch_result"`
+}
+
+// NewBatchHTTPHandler creates an instance of the Mission Control's prediction request handler
+func NewBatchHTTPHandler(mc missionctl.MissionControl) http.Handler {
+	return &batchHTTPHandler{httpHandler{mc}}
+}
+
+func (h *batchHTTPHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	var httpErr *errors.HTTPError
+	h.measureRequestDuration(httpErr)
+
+	// Create context from the request context
+	ctx := turingctx.NewTuringContext(req.Context())
+	// Create context logger
+	ctxLogger := log.WithContext(ctx)
+	defer func() {
+		_ = ctxLogger.Sync()
+	}()
+
+	// Get the unique turing request id from the context
+	turingReqID, err := turingctx.GetRequestID(ctx)
+	if err != nil {
+		ctxLogger.Errorf("Could not retrieve Turing Request ID from context: %v",
+			err.Error())
+	}
+	ctxLogger.Debugf("Received batch request for %v", turingReqID)
+
+	ctx = h.enableTracingSpan(ctx, req)
+
+	// Read the request body
+	requestBody, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		h.error(ctx, rw, errors.NewHTTPError(err))
+		return
+	}
+
+	//Valid request
+	if valid := json.Valid(requestBody); !valid {
+		h.error(ctx, rw, errors.NewHTTPError(errors.Newf(errors.BadInput,
+			`Invalid json request format`)))
+		return
+	}
+
+	//Split into batches
+	var batchRequests map[string][]interface{}
+	err = json.Unmarshal(requestBody, &batchRequests)
+	if err != nil {
+		h.error(ctx, rw, errors.NewHTTPError(errors.Newf(errors.BadInput,
+			`Invalid json request`)))
+		return
+	}
+
+	//Validate request
+	if _, ok := batchRequests["batch_request"]; !ok {
+		h.error(ctx, rw, errors.NewHTTPError(errors.Newf(errors.BadInput,
+			`batch_request" not found in request`)))
+		return
+	}
+
+	//Handle request asynchronously
+	var batchResponses = make([]batchResponse, len(batchRequests["batch_request"]))
+	var waitGroup sync.WaitGroup
+	for index, value := range batchRequests["batch_request"] {
+		waitGroup.Add(1)
+
+		go func(index int, jsonRequestBody interface{}) {
+			defer waitGroup.Done()
+			batchRequestBody, _ := json.Marshal(jsonRequestBody)
+			var batchResponse batchResponse
+			resp, httpErr := h.getPrediction(ctx, req, ctxLogger, batchRequestBody)
+			if httpErr != nil {
+				batchResponse.StatusCode = httpErr.Code
+				batchResponse.ErrorMsg = httpErr.Message
+				batchResponses[index] = batchResponse
+				return
+			}
+			err := json.Unmarshal(resp.Body(), &batchResponse.Data)
+			if err != nil {
+				batchResponse.StatusCode = http.StatusInternalServerError
+				batchResponse.ErrorMsg = "Unable to marshall response into json"
+			}
+			batchResponse.StatusCode = http.StatusOK
+			batchResponses[index] = batchResponse
+		}(index, value)
+	}
+	waitGroup.Wait()
+
+	// Write the json response to the writer
+	rw.Header().Set("Content-Type", "application/json")
```

Comment:
Maybe we can just mirror back the original request's Content-Type:
```go
rw.Header().Set("Content-Type", resp.Header().Get("Content-Type"))
```

Reply:
The example you have shown is that of the response: `resp.Header()`

> Maybe we can just mirror back the original request's Content-Type

But if you meant this:
* I am not sure if there we be cases where the request's content type header is missing.
* If there will be cases where the request content-type is json but the response type is text

So I feel it's better to copy from the response (unless Turing is crafting it, in which case it should determine the type).

Reply:
Not in favour of mirroring it, since the router does not dynamically handle all kinds of content-type.

Reply:
I do feel that copying from response is neater, but my dilemma is that if there were multiple individual request, some with failure and some with success, it might be different content type. 

In most cases, I thought that returning a json with individual responses made the most sense to me, and hence I hardcoded to it to application/json initially.

Reply:
I think the other alternative is to reject non `content-type: application/json` in requests if you want to be strict about it, then it's safe to mirror.